File: WHO all groups/Cuba 2/goveqs_basisnonHIV.m
function out = goveqs_basisnonHIV(t, in, i, s, M, agg, sel, r, p)

out = zeros(length(in),1);
invec = in(1:i.nstates);

% Prepare population denominators
tmp = M.denvec*invec;
den = sum(tmp.*M.denvec,1)';
den(den==0) = Inf;

% if t<1980                                                                  
%     rHIV = 0;
% else
%     rHIV = interp1(0:length(r.rHIV)-1, r.rHIV, (t-1980)); 
% end

try
% Normalise by populations
lam = M.lam*(invec./den)*(1-p.betadec)^(max((t-2010),0));
allmat = M.lin + lam(1)*M.nlin.ch + lam(2)*M.nlin.ad;                     
out = allmat*invec;         
catch
   keyboard; 
end


% Mortality and births
morts = M.mort.*invec;
out(1:i.nstates) = out(1:i.nstates) - sum(morts,2);

dom_morts = sum(sum(morts([s.dom,s.vuln],:)));
out(i.U.ch.dom.neg) = out(i.U.ch.dom.neg) + dom_morts;


if sum(invec)>5
    keyboard;
end


% Auxiliaries
out(i.aux.inc)        = agg.inc*(sel.inc.*allmat)*invec;
% tmp1                  = agg.incsources*((sel.L2I.*allmat)*invec);
% tmp2                  = agg.incsources*((sel.P2I.*allmat)*invec);
% tmp3                  = agg.incsources*((sel.R2I.*allmat)*invec);
% tmp4                  = agg.incsources*((sel.T2I.*allmat)*invec);
% out(i.aux.incsources) = [tmp1; tmp2; tmp3; tmp4];
out(i.aux.mort)       = sum(morts(:,2));
out(i.aux.nTPT)       = sum((sel.nTPT.*allmat)*invec);
out(i.aux.ch_notifs)  = sum((sel.ch_notifs.*allmat)*invec);

File: Python/goveqs_basisnonHIV.py


import numpy as np
from scipy.sparse import csr_matrix
import pdb
from setup_model import *

def goveqs_basisnonHIV(t, in_vec, i, s, M, agg, sel, r, p, prm):


    out = np.zeros(i['nx'])
    invec = in_vec[: i['nstates']]
    invec = np.array(invec, dtype=float)

    # Prepare population denominators
    tmp = M['denvec'] @ invec
    den = np.sum(M['denvec'].toarray() * tmp[:, None], axis=0)
    den[den == 0] = np.inf

    try:
        lam_val = (M['lam'] @ (invec / den)) * ((1 - p['betadec']) ** max(t - 2010, 0))
        allmat = M['lin'] + lam_val[0] * M['nlin']['ch'] + lam_val[1] * M['nlin']['ad']
        out = allmat.dot(invec)
    except Exception as e:
        print("Error in computing non-HIV equations:", e)
        raise

        
    # Mortality and births
    morts = M['mort'].toarray() * invec.reshape(-1, 1)
    out[: i['nstates']] = out[: i['nstates']] - np.sum(morts, axis=1)

    dom_morts = np.sum(morts[np.array(s['dom']) - 1, :])
    out[i['U']['ch']['dom']['neg'] - 1] = out[i['U']['ch']['dom']['neg'] - 1] + dom_morts
    
        # Convert arrays to numeric types
    sel_inc = np.array(sel['inc'], dtype=float)
    allmat_f = allmat.toarray()
    invec_f = np.array(invec, dtype=float)


    # Auxiliaries
    #out[np.array(i['aux']['inc']) - 1] = agg['inc'] @ (np.multiply(sel_inc, allmat_f) @ invec_f)
# Convert the auxiliary indices to integers and shift to 0-indexing
    aux_indices = np.array(i['aux']['inc'], dtype=int) - 1
    # Force any index that's too high to equal i['nx'] - 1
    aux_indices[aux_indices >= i['nx']] = i['nx'] - 1
    # Now assign
    out[aux_indices] = agg['inc'] @ (np.multiply(sel_inc, allmat_f) @ invec_f)


    out[np.array(i['aux']['mort']) - 1] = np.sum(morts[:, 1])
    out[np.array(i['aux']['nTPT']) - 1] = (sel['nTPT'].multiply(allmat) @ invec).sum()
    out[np.array(i['aux']['ch_notifs']) - 1] = (sel['ch_notifs'].multiply(allmat) @ invec).sum()


    return out


